stable_radical_optimization/stable_radical_opt.py

```diff
@@ -0,0 +1,334 @@
+            super(StableRadOptProblem, self).__init__(engine, config, **kwargs)
+            self.engine = engine
+            self._config = config
+            self.stability_model = stability_model
+            self.redox_model = redox_model
+            self.bde_model = bde_model
+
+        def get_initial_state(self) -> MoleculeState:
+            return MoleculeState(rdkit.Chem.MolFromSmiles('C'), self._config)
+
+        def get_reward(self, state: MoleculeState) -> Tuple[float, dict]:
+            # TODO what does this do? Is it needed?
+            ## Node is outside the domain of validity
+            #elif ((node.policy_inputs['atom'] == 1).any() | 
+            #    (node.policy_inputs['bond'] == 1).any()):
+            #    return config.min_reward
+
+            # TODO what data should be stored here?
+            if state.forced_terminal:
+                return self.calc_reward_inner(state), {'forced_terminal': True, 'smiles': state.smiles}
+            return 0.0, {'forced_terminal': False, 'smiles': state.smiles}
+
+        @tf.function(experimental_relax_shapes=True)                
+        def predict(self, model: 'tf.keras.Model', inputs):
+            return model.predict_step(inputs)
+
+        def calc_reward_inner(self, state: MoleculeState) -> float:
+            """
+            """
+            spins, buried_vol = self.predict(self.stability_model,
+                {key: tf.constant(np.expand_dims(val, 0))
+                for key, val in self.get_policy_inputs(state).items()})
+
+            spins = spins.numpy().flatten()
+            buried_vol = buried_vol.numpy().flatten()
+
+            atom_index = int(spins.argmax())
+            max_spin = spins[atom_index]
+            spin_buried_vol = buried_vol[atom_index]
+
+            atom_type = state.molecule.GetAtomWithIdx(atom_index).GetSymbol()
+
+            ionization_energy, electron_affinity = self.predict(
+                self.redox_model, {key: tf.constant(np.expand_dims(val, 0))
+                            for key, val in self.get_policy_inputs(state).items()})\
+                                                       .numpy().tolist()[0]
+
+            v_diff = ionization_energy - electron_affinity
+            bde, bde_diff = self.calc_bde(state)
+
+            ea_range = (-.5, 0.2)
+            ie_range = (.5, 1.2)
+            v_range = (1, 1.7)   
+            bde_range = (60, 80)    
+
+            # This is a bit of a placeholder; but the range for spin is about 1/50th that
+            # of buried volume.
+            reward = ((1 - max_spin) * 50 + spin_buried_vol 
+                    + 25 / (1 + np.exp(-(bde_diff - 10)))
+                    + 100 * (
+                self.windowed_loss(electron_affinity, ea_range) +     
+                self.windowed_loss(ionization_energy, ie_range) + 
+                self.windowed_loss(v_diff, v_range) +         
+                self.windowed_loss(bde, bde_range)) / 4)
```

Comment:
we'd probably want to store all these numbers, `max_spin`, etc., that come out of the predictive models